src/Coq.hs

```diff
@@ -0,0 +1,410 @@
+{-
+ -
+ - coq backend for act
+ -
+ - unsupported features:
+ - + bytestrings
+ - + external storage
+ - + specifications for multiple contracts
+ -
+ -}
+
+{-# Language OverloadedStrings #-}
+{-# LANGUAGE GADTs #-}
+
+module Coq where
+
+import qualified Data.Map.Strict    as M
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Text          as T
+import Data.Either (rights)
+import Data.List (find, groupBy)
+import Control.Monad.State
+
+import EVM.ABI
+import EVM.Solidity (SlotType(..))
+import Syntax
+import RefinedAst
+
+type Store = M.Map Id SlotType
+type Fresh = State Int
+
+header :: T.Text
+header =
+  "(* --- GENERATED BY ACT --- *)\n\n\
+  \Require Import Coq.ZArith.ZArith.\n\
+  \Require Import ActLib.ActLib.\n\
+  \Require Coq.Strings.String.\n\n\
+  \Module " <> strMod <> " := Coq.Strings.String.\n\
+  \Open Scope Z_scope.\n\n"
+
+-- | produce a coq representation of a specification
+coq :: Store -> [Claim] -> T.Text
+coq store claims =
+
+  header
+  <> layout store <> "\n\n"
+  <> block (evalSeq (claim store) <$> groups transitions)
+  <> block (evalSeq retVal        <$> groups transitions)
+  <> block (evalSeq (base store)  <$> groups constructors)
+  <> reachable (groups constructors) (groups transitions)
+
+  where
+
+  behaviours = [a | B a <- claims]
+
+  transitions = filter f behaviours where
+    f (Behaviour _ Pass False _ _ _ _ _ _) = True
+    f _ = False
+
+  constructors = filter f behaviours where
+    f (Behaviour _ Pass True  _ _ _ _ _ _) = True
+    f _ = False
+
+  groups = groupBy (\b b' -> _name b == _name b')
+
+  block xs = T.intercalate "\n\n" (concat xs) <> "\n\n"
+
+  layout store' =
+    "Record " <> stateType <> " : Set := " <> stateConstructor <> "\n{ "
+    <> T.intercalate ("\n" <> "; ") (map decl pairs)
+    <> "\n" <> "}." where
+    pairs = M.toList store'
+    decl (n, s) = (T.pack n) <> " : " <> slotType s
+
+
+-- | inductive definition of reachable states
+reachable :: [[Behaviour]] -> [[Behaviour]] -> T.Text
+reachable constructors groups = inductive
+
+  reachableType "" (stateType <> " -> " <> stateType <> " -> Prop") body
+
+  where
+
+  body = concat $
+    (evalSeq baseCase <$> constructors)
+    <>
+    (evalSeq reachableStep <$> groups)
+
+  reachableStep :: Behaviour -> Fresh T.Text
```

Comment:
In general, I prefer promoting functions to toplevel when they don't rely on locally scoped variables. This is the case for `reachableStep` and `baseCase`, so I think they should be promoted.

Reply:
done (https://github.com/ethereum/act/commit/dd1cacdc86e6f5dc80cda1b611fb3e8b183b7638)